[dnm] server: add multi-tenant modes for default test tenant

herkolategan opened the pull request:
Informs #106772. 
Epic: CRDB-18499

Work in progress - do not merge

Currently this PR is forced to start the default test tenant and use the shared process mode to track failures.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/99497)
<!-- Reviewable:end -->

herkolategan committed:
server: add multi-tenant modes for default test tenant

In addition to the default test tenant, the different modes a tenant can be
started in needs to be tested as well. In this change a new setting has been
added to test server args that allow selecting between a separate process,
shared process or probabilistically selecting between one of the aforementioned
modes.
